### DISCUSSION TITLE
Linker not finding objects from included library

### DISCUSSION THREAD

**Discussion by robotoshi** [nasa/fprime#2341] [Q&A]
Hello again,

I have been getting fprime set up and am very close to getting a build when cross-compiling.
I'm just having a couple linker errors when cross-compiling, where an included "library" that I've forked from the community fprime-freertos repo.  
I have just pushed everything I have to a branch in our repo.  I apologize for the mess, there may be some files in there that don't need to be, I just included everything I'm working with right now.  This is a rough work in progress.

I've tried to make the project more "moudular" using included submodules, but I think I'm just not sure how to set up cmake properly to do that.
I've tried including files in MOD_DEPS, but I'm kind of confused exactly how it needs to be done to make it work.

If you have any other suggestions or advice for restructuring the project, I am open.

I have been using the `arm-none-eabi-g++` compiler, but I don't think it matters too much since this is a cmake issue.

The repo at the branch I'm sharing is https://github.com/CalgaryToSpace/CTS-SAT-1-Software/tree/wip
It includes a submodule https://github.com/CalgaryToSpace/fprime-freertos/tree/wip which I am also actively working on.  I plan to separate out the interdependencies and share this repo for others to use eventually.  This is where the changes need to be made to fix my linker problem.

When I try to build the Nucleo deployment, these are some of the linker errors I get.  All the reset are similar.

Command:
```
env PATH="${PATH}:~/cubesat/CTSSAT1/arm-toolchain/" fprime-utils generate stm32
cd Nucleo/
fprime-utils build stm32
```
Errors:

```
ld: ../lib/stm32/libNucleo_Top.a(NucleoTopology.cpp.obj): in function `Os::Log::~Log()':
NucleoTopology.cpp:(.text._ZN2Os3LogD2Ev[_ZN2Os3LogD5Ev]+0x20): undefined reference to `vtable for Os::Log'

ld: ../lib/stm32/libNucleo_Top.a(NucleoTopology.cpp.obj): in function `__static_initialization_and_destruction_0(int, int)':
NucleoTopology.cpp:(.text._Z41__static_initialization_and_destruction_0ii+0x20): undefined reference to `Os::Log::Log()'

ld: NucleoTopology.cpp:(.text._Z41__static_initialization_and_destruction_0ii+0x18c): undefined reference to `Os::Mutex::Mutex()'

ld: NucleoTopology.cpp:(.text._Z41__static_initialization_and_destruction_0ii+0x208): undefined reference to `Os::Mutex::~Mutex()'

ld: ../lib/stm32/libNucleo_Drv_STM32UartDriver.a(STM32UartDriver.cpp.obj): in function `Drv::STM32UartDriver::STM32UartDriver(char const*)':
STM32UartDriver.cpp:(.text._ZN3Drv15STM32UartDriverC2EPKc+0x34): undefined reference to `Os::Task::Task()'

ld: ../lib/stm32/libNucleo_Drv_STM32UartDriver.a(STM32UartDriver.cpp.obj): in function `Drv::STM32UartDriver::~STM32UartDriver()':
STM32UartDriver.cpp:(.text._ZN3Drv15STM32UartDriverD2Ev+0x16): undefined reference to `HAL_UART_DeInit'

ld: STM32UartDriver.cpp:(.text._ZN3Drv15STM32UartDriverD2Ev+0x22): undefined reference to `Os::Task::~Task()'

ld: ../lib/stm32/libNucleo_Drv_STM32UartDriver.a(STM32UartDriver.cpp.obj): in function `Drv::STM32UartDriver::send_handler(long, Fw::Buffer&)':
STM32UartDriver.cpp:(.text._ZN3Drv15STM32UartDriver12send_handlerElRN2Fw6BufferE+0x6c): undefined reference to `HAL_UART_Transmit'

```

Thank you in advance for your help

**Comment by manas-narayan** [comment]
Hey @robotoshi did you get fprime working with FreeRTOS and deployed it on the STM board

**Comment by thomas-bc** [comment]
FYI we now have a guide on how to work with external libraries
https://fprime.jpl.nasa.gov/devel/docs/how-to/integrate-external-libraries/
